### ISSUE TITLE
TestKVHash fail

### ISSUE THREAD

**Issue by yichengq** [coreos/etcd#3560]
```
--- FAIL: TestKVHash (0.16s)
	kv_test.go:642: hash[0](1031971881) != hash[1](1321773765)
	kv_test.go:642: hash[1](1321773765) != hash[2](1031971881)
```

HEAD at 09/18 06180be1540, go1.4

**Comment by gyuho** [comment]
I am getting the same crash.

```
--- FAIL: TestKVHash-16 (0.78s)
	kv_test.go:646: hash[0](1321773765) != hash[1](1031971881)
	kv_test.go:646: hash[1](1031971881) != hash[2](1321773765)
```

From the command `go test -bench . -benchmem -cpu 1,2,4,8,16`
and run on the master branch.

https://github.com/coreos/etcd/commit/3b70bf87c3f8b8d8322dbe55226c8af80da21a9f
And Go1.5.

Thanks,

**Comment by gyuho** [comment]
The hash function never fails as far as I know in normal test.

It only fails when I run benchmark.

I did the same benchmark on same hash function https://github.com/coreos/etcd/blob/master/storage/kvstore.go#L291
but without `Snapshot`, and got no crash.

So maybe something related with Snapshot code here:
https://github.com/coreos/etcd/blob/master/storage/backend/backend.go#L87-L99
?

I want to try again to reproduce the failure tonight but any suggestion to debug this?

Thanks,

**Comment by yichengq** [comment]
I have no clue about it too.

**Comment by yichengq** [comment]
I could hit this in travis in low probability.

It could read out the keys put, so hash difference doesn't affect the correctness.

**Comment by gyuho** [comment]
Yeah, I also confirmed that raw data from Snapshot to boltdb was identical.
It just rarely has different hash values, in most cases when you run benchmarks
with more than one CPU.

Not sure why this only happens in benchmarks, as far as I observed so far.
Tests running in parallel does not crash from Hash. Only crash in benchmarks--for me.

I will investigate more.

Thanks,

**Comment by gyuho** [comment]
Never mind. I think @xiang90  figured it out with https://github.com/coreos/etcd/pull/3585.

Thanks all!

**Comment by xiang90** [comment]
Fixed by #3585
